### PR TITLE
Entity editor links table design & functionality

### PR DESCRIPTION
**Pull request by yusufkinatas**
## 🌟 What is the purpose of this PR?

This PR introduces updates the design of the links table, and adds the following functionality to links table
- remove link
- add link
- change link

Note that links with `maxItems > 1` is not supported yet. So only single links are supported, not link arrays. There will be follow-up PRs. 

## 🔗 Related links

- [Design](https://www.figma.com/file/gydVGka9FjNEg9E2STwhi2/HASH-App?node-id=6468%3A253538&t=5aqL5g0lLyxTzeyF-4) _(internal)_

## 🔍 What does this change?

- 

## ⚠️ Known issues

- [periodic ‘must have right endpoint’ crash when repeatedly archiving and creating links to replace a link](https://hashintel.slack.com/archives/C022217GAHF/p1669893629364659?thread_ts=1669891299.452099&cid=C022217GAHF) _(internal)_

## 🐾 Next steps

Implement list of links when `maxItems > 1` on a link

## ❓ How to test this?

1.  Checkout the branch / view the deployment
1.  Create an entity with links (you can use `create demo entity type` button here https://github.com/hashintel/hash/commit/bfa2f02480b29584944c48ced897a9c3ecd92458)
1.  Confirm that entity editor links table has the functionality shown in the video below at demo section

## 📹 Demo

https://user-images.githubusercontent.com/39553853/205078310-8e332afa-a886-495a-b888-31b815b5d1c4.mov

### CODE REVIEW

File: packages/hash/frontend/src/pages/[account-slug]/entities/[entity-uuid].page.tsx [hunk 1]
```diff
@@ -50,6 +55,16 @@ const Page: NextPageWithLayout = () => {
             } catch {
               setEntitySubgraph(undefined);
             }
+
+            const entityTypeId = getRoots(subgraph)[0]?.metadata.entityTypeId;
+
+            if (entityTypeId) {
+              const { data: typeSubgraph } = await getEntityType({
```
Comment (nathggns) [inline packages/hash/frontend/src/pages/[account-slug]/entities/[entity-uuid].page.tsx:62]: Curious why we're having to do an extra roundtrip for this – can this not be resolved in the subgraph of `getEntity`?

File: packages/hash/frontend/src/pages/[account-slug]/entities/[entity-uuid].page/entity-editor/links-section/link-table/cells/linked-with-cell.tsx [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+    const yCenter = getYCenter(args);
+    const cellPadding = getCellHorizontalPadding();
+    const left = rect.x + cellPadding;
+
+    const linkAndTargetEntity = linkAndTargetEntities[0];
+
+    if (!linkAndTargetEntity) {
+      // draw empty value
+      ctx.fillStyle = "#91A5BA";
```
Comment (nathggns) [inline packages/hash/frontend/src/pages/[account-slug]/entities/[entity-uuid].page/entity-editor/links-section/link-table/cells/linked-with-cell.tsx:46]: Should this come from the theme?

File: packages/hash/frontend/src/pages/[account-slug]/entities/[entity-uuid].page/entity-editor/links-section/link-table/cells/linked-with-cell/linked-with-cell-editor.tsx [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+
+  const onCreateNew = () => {
+    if (!expectedEntityTypes[0]) {
+      return;
+    }
+
+    const accountSlug = router.query["account-slug"];
+    /** @todo ask about this behavior */
+    void window.open(
+      `/${accountSlug}/entities/new?entity-type-id=${encodeURIComponent(
```
Comment (nathggns) [inline packages/hash/frontend/src/pages/[account-slug]/entities/[entity-uuid].page/entity-editor/links-section/link-table/cells/linked-with-cell/linked-with-cell-editor.tsx:65]: Think this URL is now `@shortname/new/entity`

File: packages/hash/frontend/src/pages/[account-slug]/entities/[entity-uuid].page/entity-editor/links-section/link-table/cells/linked-with-cell/linked-with-cell-editor.tsx [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+    void init();
+  }, [aggregateEntities, expectedEntityTypes]);
+
+  const onCreateNew = () => {
+    if (!expectedEntityTypes[0]) {
+      return;
+    }
+
+    const accountSlug = router.query["account-slug"];
+    /** @todo ask about this behavior */
```
Comment (nathggns) [inline packages/hash/frontend/src/pages/[account-slug]/entities/[entity-uuid].page/entity-editor/links-section/link-table/cells/linked-with-cell/linked-with-cell-editor.tsx:63]: I imagine eventually we want a new entity modal?

File: packages/hash/frontend/src/pages/[account-slug]/entities/[entity-uuid].page/entity-editor/links-section/link-table/cells/linked-with-cell/linked-with-cell-editor.tsx [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+          },
+        },
+        variant: "entity",
+      }}
+      loading={loading}
+      options={entities.sort((a, b) =>
+        a.metadata.editionId.baseId.localeCompare(b.metadata.editionId.baseId),
+      )}
+      optionToRenderData={(entity) => ({
+        /** @todo $id is wrong */
```
Comment (nathggns) [inline packages/hash/frontend/src/pages/[account-slug]/entities/[entity-uuid].page/entity-editor/links-section/link-table/cells/linked-with-cell/linked-with-cell-editor.tsx:132]: ?

File: packages/hash/frontend/src/pages/[account-slug]/entities/[entity-uuid].page/entity-editor/links-section/link-table/cells/linked-with-cell/linked-with-cell-editor.tsx [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+          onMouseDown: (evt) => {
+            evt.preventDefault();
+            evt.stopPropagation();
+            onCreateNew();
+          },
+        },
+        variant: "entity",
+      }}
+      loading={loading}
+      options={entities.sort((a, b) =>
```
Comment (nathggns) [inline packages/hash/frontend/src/pages/[account-slug]/entities/[entity-uuid].page/entity-editor/links-section/link-table/cells/linked-with-cell/linked-with-cell-editor.tsx:128]: `.sort` mutates, and `entities` is in React state. You need to wrap this with `[...entities]`

File: packages/hash/frontend/src/pages/[account-slug]/types/entity-type/hash-selector-autocomplete.tsx [hunk 7]
```diff
@@ -133,7 +133,12 @@ export const HashSelectorAutocomplete = <T extends unknown>({
 
         // @todo extract component
         return (
-          <li {...props} data-testid="propery-selector-option">
+          <li
+            {...props}
+            data-testid="property-selector-option"
+            /** added "click-outside-ignore" to be able to use this selector with Grid component */
```
[the remark this answers is not in the document]
Comment (yusufkinatas) [reply packages/hash/frontend/src/pages/[account-slug]/types/entity-type/hash-selector-autocomplete.tsx:139]: Elements with "click-outside-ignore" classname are not closing the grid editor overlay when clicked. This behavior comes from glide-grid itself. See it here 
https://github.com/glideapps/glide-data-grid/blob/main/packages/core/src/click-outside-container/click-outside-container.tsx#L23-L25

File: packages/hash/frontend/src/pages/[account-slug]/entities/new.page.tsx [hunk 8]
```diff
@@ -13,13 +13,13 @@ const Page: NextPageWithLayout = () => {
   const createNewEntityAndRedirect = useCreateNewEntityAndRedirect();
   const [loading, setLoading] = useState(true);
 
-  const entityTypeId = String(router.query["entity-type-id"]);
```
[the remark this answers is not in the document]
Comment (yusufkinatas) [reply packages/hash/frontend/src/pages/[account-slug]/entities/new.page.tsx]: This page is removed now.

File: packages/hash/frontend/src/pages/[account-slug]/entities/[entity-uuid].page/entity-editor/links-section/link-table/use-create-get-cell-content.ts [hunk 9]
```diff
@@ -19,14 +23,62 @@ export const useCreateGetCellContent = () => {
...
+              kind: GridCellKind.Custom,
+              readonly: true,
+              allowOverlay: false,
+              copyData: String(linkRow.linkTitle),
+              data: {
+                kind: "link-cell",
+                linkRow,
+              } as LinkCellProps,
+            };
+          case "linkedWith":
```
[the remark this answers is not in the document]
Comment (yusufkinatas) [reply packages/hash/frontend/src/pages/[account-slug]/entities/[entity-uuid].page/entity-editor/links-section/link-table/use-create-get-cell-content.ts:38]: `secondaryText` and `count` variables here are not conflicting with any other variable here. Typescript throws an error if that happens. When that's the case I guess it's optional to use braces or not. Let me know if there is a common practice on this on the repository, or another concern. Maybe we can discuss an eslint rule for this if you want.

File: packages/hash/frontend/src/pages/[account-slug]/entities/[entity-uuid].page/entity-editor/links-section/link-table/use-create-get-cell-content.ts [hunk 10]
```diff
@@ -19,14 +23,62 @@ export const useCreateGetCellContent = () => {
           throw new Error("columnKey not found");
         }
 
-        const value = link[columnKey];
+        switch (columnKey) {
```
[the remark this answers is not in the document]
Comment (yusufkinatas) [reply packages/hash/frontend/src/pages/[account-slug]/entities/[entity-uuid].page/entity-editor/links-section/link-table/use-create-get-cell-content.ts:26]: If you mean `kind: "link-cell"` => `kind: "link-cell" as const`, that does not help. Or maybe I got you wrong, let me know.

File: packages/hash/frontend/src/pages/[account-slug]/entities/[entity-uuid].page/entity-editor/links-section/link-table/use-create-get-cell-content.ts [hunk 11]
```diff
@@ -19,14 +23,62 @@ export const useCreateGetCellContent = () => {
...
+                kind: "linked-with-cell",
+                linkRow,
+              } as LinkedWithCellProps,
+            };
+          case "expectedEntityTypes":
+            return {
+              kind: GridCellKind.Custom,
+              readonly: true,
+              allowOverlay: true,
+              copyData: String(linkRow.expectedEntityTypes),
```
[the remark this answers is not in the document]
Comment (yusufkinatas) [reply packages/hash/frontend/src/pages/[account-slug]/entities/[entity-uuid].page/entity-editor/links-section/link-table/use-create-get-cell-content.ts:74]: Oops, it should be `expectedEntityTypes`, which is a string array. `String(["1", "2"])` becomes `"1,2"`. Which should work as expected now.

File: packages/hash/frontend/src/pages/[account-slug]/entities/[entity-uuid].page/entity-editor/links-section/link-table/cells/linked-with-cell/linked-with-cell-editor.tsx (hunk 6 above)
Comment (yusufkinatas) [reply packages/hash/frontend/src/pages/[account-slug]/entities/[entity-uuid].page/entity-editor/links-section/link-table/cells/linked-with-cell/linked-with-cell-editor.tsx:128]: Oops missed this, thanks!

File: packages/hash/frontend/src/pages/[account-slug]/entities/[entity-uuid].page.tsx (hunk 1 above)
Comment (yusufkinatas) [reply packages/hash/frontend/src/pages/[account-slug]/entities/[entity-uuid].page.tsx:62]: Wait, you're totally right. The whole `entityTypeSubgraph` is unnecessary 🤦‍♂️ . Just got ride of all. Thanks for noticing!

**Comment by yusufkinatas** [comment]
Thanks for the review, addressed to all comments above!

File: packages/hash/frontend/src/pages/[account-slug]/entities/[entity-uuid].page/entity-editor/links-section/link-table/cells/linked-with-cell/linked-with-cell-editor.tsx (hunk 5 above)
Comment (yusufkinatas) [reply packages/hash/frontend/src/pages/[account-slug]/entities/[entity-uuid].page/entity-editor/links-section/link-table/cells/linked-with-cell/linked-with-cell-editor.tsx]: Updated the comment

File: packages/hash/frontend/src/pages/[account-slug]/entities/[entity-uuid].page/entity-editor/links-section/link-table/cells/linked-with-cell/linked-with-cell-editor.tsx (hunk 3 above)
Comment (yusufkinatas) [reply packages/hash/frontend/src/pages/[account-slug]/entities/[entity-uuid].page/entity-editor/links-section/link-table/cells/linked-with-cell/linked-with-cell-editor.tsx]: Correct, updated the URL

File: packages/hash/frontend/src/pages/[account-slug]/entities/[entity-uuid].page/entity-editor/links-section/link-table/cells/linked-with-cell.tsx (hunk 2 above)
Comment (yusufkinatas) [reply packages/hash/frontend/src/pages/[account-slug]/entities/[entity-uuid].page/entity-editor/links-section/link-table/cells/linked-with-cell.tsx]: Yes, ideally, used `customColors` object for now, but we should eventually use the theme (to support switching themes)

File: packages/hash/frontend/src/pages/[account-slug]/entities/[entity-uuid].page/entity-editor/properties-section/property-table/cells/chip-cell.tsx [hunk 12]
```diff
@@ -4,39 +4,63 @@ import {
...
+  color?: ChipProps["color"];
 }
 
 export type ChipCell = CustomCell<ChipCellProps>;
 
+const getChipColors = (color: ChipProps["color"]) => {
+  switch (color) {
+    case "blue":
+      /** @todo we should use colors from MUI theme */
+      return { textColor: "#0775E3", bgColor: "#E0F4FF" };
```
[the remark this answers is not in the document]
Comment (yusufkinatas) [reply packages/hash/frontend/src/pages/[account-slug]/entities/[entity-uuid].page/entity-editor/properties-section/property-table/cells/chip-cell.tsx]: Updated them here https://github.com/hashintel/hash/pull/1540/commits/5a81ec7525871dfc9d6b28a8dfaee60997f82440

File: packages/hash/frontend/src/pages/[account-slug]/entities/[entity-uuid].page/entity-editor/links-section/link-table/use-create-get-cell-content.ts [hunk 13]
```diff
@@ -19,14 +23,62 @@ export const useCreateGetCellContent = () => {
...
         }
 
-        const value = link[columnKey];
+        switch (columnKey) {
+          case "linkTitle":
+            return {
+              kind: GridCellKind.Custom,
+              readonly: true,
+              allowOverlay: false,
+              copyData: String(linkRow.linkTitle),
```
[the remark this answers is not in the document]
Comment (yusufkinatas) [reply packages/hash/frontend/src/pages/[account-slug]/entities/[entity-uuid].page/entity-editor/links-section/link-table/use-create-get-cell-content.ts]: Right, updated here https://github.com/hashintel/hash/pull/1540/commits/94b3dcff8f8a66afc6981f62c7de9010526c6ae0

File: packages/hash/frontend/src/pages/[account-slug]/entities/[entity-uuid].page/entity-editor/links-section/link-table/use-row-data.ts [hunk 14]
```diff
@@ -2,45 +2,82 @@ import { useMemo } from "react";
...
+        entitySubgraph,
+        deleteLink: async (linkEntityId) => {
+          try {
+            await archiveEntity({ data: { entityId: linkEntityId } });
+            await refetch();
+          } catch {
+            snackbar.error("Failed to remove link");
+          }
+        },
+      } as LinkRow;
```
[the remark this answers is not in the document]
Comment (yusufkinatas) [reply packages/hash/frontend/src/pages/[account-slug]/entities/[entity-uuid].page/entity-editor/links-section/link-table/use-row-data.ts]: Updated here, updated here https://github.com/hashintel/hash/pull/1540/commits/94b3dcff8f8a66afc6981f62c7de9010526c6ae0

File: packages/hash/frontend/src/pages/[account-slug]/entities/[entity-uuid].page/entity-editor/links-section/link-table/use-row-data.ts [hunk 15]
```diff
@@ -2,45 +2,82 @@ import { useMemo } from "react";
...
+    const linkSchemas = entityType.schema.links;
+
+    if (!linkSchemas) {
       return [];
     }
 
-    const entity = getRoots(entitySubgraph)[0]!;
+    return Object.keys(linkSchemas).map((key) => {
+      const linkEntityTypeId = key as VersionedUri;
+      const linkSchema = linkSchemas[linkEntityTypeId]!;
```
[the remark this answers is not in the document]
Comment (yusufkinatas) [reply packages/hash/frontend/src/pages/[account-slug]/entities/[entity-uuid].page/entity-editor/links-section/link-table/use-row-data.ts]: Thanks, updated here https://github.com/hashintel/hash/pull/1540/commits/94b3dcff8f8a66afc6981f62c7de9010526c6ae0

File: packages/hash/frontend/src/pages/[account-slug]/types/entity-type/hash-selector-autocomplete.tsx [hunk 16]
```diff
@@ -133,7 +133,12 @@ export const HashSelectorAutocomplete = <T extends unknown>({
...
+            className={`${props.className} click-outside-ignore`}
```
[the remark this answers is not in the document]
Comment (yusufkinatas) [reply packages/hash/frontend/src/pages/[account-slug]/types/entity-type/hash-selector-autocomplete.tsx]: Makes sense, updated here https://github.com/hashintel/hash/pull/1540/commits/94b3dcff8f8a66afc6981f62c7de9010526c6ae0

**Review by CiaranMn — APPROVED** [review]

**Pull request merged**